<h1 align=center><font size = 5>Segmenting and Clustering Neighborhoods in Toronto</font></h1>

## Introduction

This notebook will convert addresses into their equivalent latitude and longitude values. This notebook will use the Foursquare API to explore neighborhoods in Toronto.  The Foursquare *explore* function is used to get the most common venue categories in each neighborhood, and then this feature is used to group the neighborhoods into clusters. The _k_-means clustering algorithm will also be used to help complete this task. Finally, the Folium library will be used to visualize the neighborhoods in Toronto and their emerging clusters.

---

## 1. Download and Explore Toronto Postal Codes Dataset

In [1]:
import pandas as pd

Data was scraped from the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe.

In [2]:
data = pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")

**data** returned three lists; with the Postal Code data in the first list.

In [3]:
type(data)

list

In [4]:
len(data)

3

In [5]:
data[0]

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [6]:
df = data[0]  # transform first list to dataframe
df.head(12)   # inspect first 11 rows

Postal Code           Borough                                Neighbourhood
0          M1A      Not assigned                                 Not assigned
1          M2A      Not assigned                                 Not assigned
2          M3A        North York                                    Parkwoods
3          M4A        North York                             Victoria Village
4          M5A  Downtown Toronto                    Regent Park, Harbourfront
5          M6A        North York             Lawrence Manor, Lawrence Heights
6          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
7          M8A      Not assigned                                 Not assigned
8          M9A         Etobicoke      Islington Avenue, Humber Valley Village
9          M1B       Scarborough                               Malvern, Rouge
10         M2B      Not assigned                                 Not assigned
11         M3B        North York                                    Don Mills

### Data Cleaning needs:

    1. Column headers changed to: PostalCode, Borough, and Neighborhood
    2. Ignore cells with **Not assigned** in Borough; I'm assuming that means remove them
    3. Cell with **Not assigned** in Neighborhood, set equal to Borough 
    4. Verify Unique Postal Codes
    

In [7]:
df.columns = ['PostalCode', 'Borough', 'Neighborhood']
df = df[df.Borough != 'Not assigned']
df = df.apply(lambda x: x.Borough if x.Neighborhood == 'Not assigned' else x, axis=1)
df

PostalCode           Borough  \
2          M3A        North York   
3          M4A        North York   
4          M5A  Downtown Toronto   
5          M6A        North York   
6          M7A  Downtown Toronto   
..         ...               ...   
160        M8X         Etobicoke   
165        M4Y  Downtown Toronto   
168        M7Y      East Toronto   
169        M8Y         Etobicoke   
178        M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [8]:
# Verify all Borough cells have an assigned value
df.loc[df['Borough'] == 'Not assigned'].shape

(0, 3)

Verify there are unique Postal Codes

In [9]:
print('The dataframe has {} unique Postal Codes and {} unique Neighborhoods.'.format(
        len(df['PostalCode'].unique()),
        len(df['Neighborhood'].unique())
    )
)
print ("\nVerify Duplicate Neighborhoods have unique Postal Codes:")
print (df[df.duplicated(subset=['Neighborhood'])])

The dataframe has 103 unique Postal Codes and 99 unique Neighborhoods.

Verify Duplicate Neighborhoods have unique Postal Codes:
   PostalCode     Borough Neighborhood
20        M3C  North York    Don Mills
74        M3L  North York    Downsview
83        M3M  North York    Downsview
92        M3N  North York    Downsview


### Size of the resulting Dataframe after data cleansing

In [10]:
df.shape

(103, 3)

---

---

## 2. Retrieve Geo-location Data for Toronto Neighborhoods

Add libraries 

In [145]:
!pip install geocoder
import geocoder # import geocoder

     |████████████████████████████████| 102kB 8.5MB/s ta 0:00:011
     |████████████████████████████████| 829kB 15.0MB/s eta 0:00:01
  Using cached https://files.pythonhosted.org/packages/d2/3d/fa76db83bf75c4f8d338c2fd15c8d33fdd7ad23a9b5e57eb6c5de26b430e/click-7.1.2-py2.py3-none-any.whl
  Stored in directory: /home/jupyterlab/.cache/pip/wheels/8b/99/a0/81daf51dcd359a9377b110a8a886b3895921802d2fc1b2397e
Successfully built future


Add Latitude and Longitude Columns to the Dataframe

In [12]:
df = df.assign(Latitude='', Longitude='')

Find and update Latitude and Longitude for all Postal Codes

In [146]:
# Loop for all rows in the Dataframe
for ind in df.index:
    lat_lng_coords = None
    i = 0
    print(df['PostalCode'][ind])
    # loop until you get the coordinates
    while(lat_lng_coords is None and i < 2):
          g = geocoder.google('{}, Toronto, Ontario'.format(df['PostalCode'][ind]))
          print(g)
          lat_lng_coords = g.latlng
          i= i + 1
            
    if lat_lng_coords is not None:
        df['Latitude'][ind] = lat_lng_coords[0]
        df['longitude'][ind] = lat_lng_coords[1]
    print('Postal Code {} unique Postal Codes and {} unique Neighborhoods.'.format(
        df['PostalCode'][ind],df['Latitude'][ind], df['Longitude'][ind]))

M3A
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
Postal Code M3A unique Postal Codes and 43.7532586 unique Neighborhoods.
M4A
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
Postal Code M4A unique Postal Codes and 43.725882299999995 unique Neighborhoods.
M5A
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
Postal Code M5A unique Postal Codes and 43.6542599 unique Neighborhoods.
M6A
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
Postal Code M6A unique Postal Codes and 43.718517999999996 unique Neighborhoods.
M7A
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
Postal Code M7A unique Postal Codes and 43.6623015 unique Neighborhoods.
M9A
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
Postal Code M9A unique Postal Codes and 43.6678556 unique Neighborh

The above code executed for all 103 postal codes, but all returned the same result:

In [15]:
<[REQUEST_DENIED] Google - Geocode [empty]>

SyntaxError: invalid syntax (<ipython-input-15-29dbc764616a>, line 1)

So, my assumption, if I want to finish this assignment, I'll have to read the CSV file for the Lat/Lon

In [16]:
geo_df = pd.read_csv("Geospatial_Coordinates.csv")
geo_df.columns = ['PostalCode', 'Latitude', 'Longitude']

In [17]:
geo_df.tail(25)

PostalCode   Latitude  Longitude
78         M6K  43.636847 -79.428191
79         M6L  43.713756 -79.490074
80         M6M  43.691116 -79.476013
81         M6N  43.673185 -79.487262
82         M6P  43.661608 -79.464763
83         M6R  43.648960 -79.456325
84         M6S  43.651571 -79.484450
85         M7A  43.662301 -79.389494
86         M7R  43.636966 -79.615819
87         M7Y  43.662744 -79.321558
88         M8V  43.605647 -79.501321
89         M8W  43.602414 -79.543484
90         M8X  43.653654 -79.506944
91         M8Y  43.636258 -79.498509
92         M8Z  43.628841 -79.520999
93         M9A  43.667856 -79.532242
94         M9B  43.650943 -79.554724
95         M9C  43.643515 -79.577201
96         M9L  43.756303 -79.565963
97         M9M  43.724766 -79.532242
98         M9N  43.706876 -79.518188
99         M9P  43.696319 -79.532242
100        M9R  43.688905 -79.554724
101        M9V  43.739416 -79.588437
102        M9W  43.706748 -79.594054

Merged both dataframes with matching on the PostalCode value; cleaned up the resulting dataframe

In [18]:
df=df.merge(geo_df,on='PostalCode',how='outer')
df = df.drop(columns=['Latitude_x', 'Longitude_x'])
df = df.rename({'Latitude_y': 'Latitude', 'Longitude_y': 'Longitude'}, axis=1) 

Verified several rows from the geo dataframe matched up correctly with the resulting table.

In [19]:
print(geo_df[geo_df.PostalCode == 'M5G'])
print(df[df.PostalCode == 'M5G'])
print(geo_df[geo_df.PostalCode == 'M6M'])
print(df[df.PostalCode == 'M6M'])
print(geo_df[geo_df.PostalCode == 'M9A'])
print(df[df.PostalCode == 'M9A'])

   PostalCode   Latitude  Longitude
57        M5G  43.657952 -79.387383
   PostalCode           Borough        Neighborhood   Latitude  Longitude
24        M5G  Downtown Toronto  Central Bay Street  43.657952 -79.387383
   PostalCode   Latitude  Longitude
80        M6M  43.691116 -79.476013
   PostalCode Borough                                      Neighborhood  \
56        M6M    York  Del Ray, Mount Dennis, Keelsdale and Silverthorn   

     Latitude  Longitude  
56  43.691116 -79.476013  
   PostalCode   Latitude  Longitude
93        M9A  43.667856 -79.532242
  PostalCode    Borough                             Neighborhood   Latitude  \
5        M9A  Etobicoke  Islington Avenue, Humber Valley Village  43.667856   

   Longitude  
5 -79.532242  


In [20]:

df.head(12)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   

                                         Neighborhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
5             Islington Avenue, Humber Valley Village  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                           Don Mills  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724

---

---

## 3. Explore and cluster the neighborhoods in Toronto

#### Load libraries for visualization

In [21]:
import numpy as np # library to handle data in a vectorized manner

#import requests # library to handle requests
#from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

neighborhoods = df

**Create a map of Toronto with neighborhoods superimposed on top.**

In [149]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map([43.6534817, -79.3839347], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Segment only the neighborhoods with the word *Toronto* in the Borough; 
this filter was selected because the number of neighborhoods are around 40 (not too big and not too small)

In [40]:
toronto_data = neighborhoods[neighborhoods['Borough'].str.contains("Toronto")]
print(toronto_data.shape)
toronto_data.head(10)

(39, 5)


PostalCode           Borough                                 Neighborhood  \
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
9         M5B  Downtown Toronto                     Garden District, Ryerson   
15        M5C  Downtown Toronto                               St. James Town   
19        M4E      East Toronto                                  The Beaches   
20        M5E  Downtown Toronto                                  Berczy Park   
24        M5G  Downtown Toronto                           Central Bay Street   
25        M6G  Downtown Toronto                                     Christie   
30        M5H  Downtown Toronto                     Richmond, Adelaide, King   
31        M6H      West Toronto                 Dufferin, Dovercourt Village   

     Latitude  Longitude  
2   43.654260 -79.360636  
4   43.662301 -79.389494  
9   43.657162 -79.378937  
15  43.651494 -79.375418  
19  43.676357 -79.293031  
20  43.644771 -79.373306  
24  43.657952 -79.387383  
25  43.669542 -79.422564  
30  43.650571 -79.384568  
31  43.669005 -79.442259

Find coordinates for Toronto

In [35]:
!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim

Solving environment: done

## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs:
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    geopy-2.0.0                |     pyh9f0ad1d_0          63 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          97 KB

The following NEW packages will be INSTALLED:

  geographiclib      conda-forge/noarch::geographiclib-1.50-py_0
  geopy              conda-forge/noarch::geopy-2.0.0-pyh9f0ad1d_0



geopy-2.0.0          | 63 KB     | ##################################### | 100% 
geographiclib-1.50   | 34 KB     | ##################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: d

In [36]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="ca_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


**Visualize the neighborhoods in Toronto with the zoom_start set to 12**

In [43]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Define Foursquare Credentials and Version (credentials were changed for security reasons)

In [62]:
CLIENT_ID = '****************************' # your Foursquare ID
CLIENT_SECRET = '*************************************' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquaredd API limit value

Find a sample neighborhood, I chose *St. James Town* because it only has one neighborhood listed

In [53]:
toronto_data.loc[15, 'Neighborhood']

'St. James Town'

Get the neighborhood's latitude and longitude values.

In [54]:
neighborhood_latitude = toronto_data.loc[15, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto_data.loc[15, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto_data.loc[15, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of St. James Town are 43.6514939, -79.3754179.


Get the top 100 venues that are in St. James Town within a radius of 500 meters.

In [60]:
# type your answer here
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
#url # display URL  hidden for security purposes


**Make FourSquare API request**

In [61]:
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5faa338c68a1837662b6077b'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'St. Lawrence',
  'headerFullLocation': 'St. Lawrence, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 85,
  'suggestedBounds': {'ne': {'lat': 43.6559939045, 'lng': -79.36921018606671},
   'sw': {'lat': 43.646993895499996, 'lng': -79.3816256139333}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '574ad72238fa943556d93b8e',
       'name': 'Gyu-Kaku Japanese BBQ',
       'location': {'address': '81 Church St',
        'crossStreet': 'at Adelaide St E',
        'lat': 43.651422275497914,
        'lng': -79.37504693687086,
        'labeledLatLngs':

Cleanup Results

In [63]:
# function from FOURSQUARE that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [64]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name           categories        lat        lng
0  Gyu-Kaku Japanese BBQ  Japanese Restaurant  43.651422 -79.375047
1                Terroni   Italian Restaurant  43.650927 -79.375602
2      Fahrenheit Coffee          Coffee Shop  43.652384 -79.372719
3            Pearl Diver            Gastropub  43.651481 -79.373600
4      GEORGE Restaurant           Restaurant  43.653346 -79.374445

### Venues were returned by Foursquare

In [65]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

85 venues were returned by Foursquare.


## Explore Neighborhoods in Toronto

#### Create a function to repeat the same process to all the neighborhoods in Toronto

In [67]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Execute the above function with all Toronto neighborhoods

In [69]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

Verify data and size

In [70]:
print(toronto_venues.shape)
toronto_venues.head()

(1624, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4          Impact Kitchen       43.656369       -79.356980   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4           Restaurant

### Find out how many venues were returned for each neighborhood

In [71]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            55   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     16   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Central Bay Street                                                     68   
Christie                                                               16   
Church and Wellesley                                                   75   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             33   
Davisville North                                                        9   
Dufferin, Dovercourt Village                                           13   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          25   
India Bazaar, The Beaches West                                         19   
Kensington Market, Chinatown, Grange Park                              74   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               45   
Moore Park, Summerhill East                                             2   
North Toronto West, Lawrence Park                                      18   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            33   
Regent Park, Harbourfront                                              44   
Richmond, Adelaide, King                                              100   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     33   
St. James Town                                                         85   
St. James Town, Cabbagetown                                            48   
Stn A PO Boxes                                                         96   
Studio District                                                        37   
Summerhill West, Rathnelly, South Hill, Forest ...                     14   
The Annex, North Midtown, Yorkville                                    19   
The Beaches                                                             4   
The Danforth West, Riverdale                                           43   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             55   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre, South Ce...                      16   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      68   
Christie                                                                16   
Church and Wellesley                                                    75   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              33   
Davisville North                    

In [72]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 237 uniques categories.


## Analyze Each Neighborhood

In [73]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  ...  Theater  Theme Restaurant  Toy / Game Store  Trail  \
0             0  ...        0                 0                 0      0   
1             0  ...        0                 0                 0      0   
2             0  ...        0                 0                 0      0   
3             0  ...        0                 0                 0      0   
4             0  ...        0                 0                 0      0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0              0                              0                 0   
1              0                              0                 0   
2              0                              0                 0   
3              0                              0                 0   
4              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Women's Store  
0                      0         0              0  
1                      0         0              0  
2                      0         0              0  
3                      0         0              0  
4                      0         0              0  

[5 rows x 237 columns]

 **Examine the new dataframe size**

In [74]:
toronto_onehot.shape

(1624, 237)

**Group rows by neighborhood and by taking the mean of the frequency of occurrence of each category**

In [75]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1        Brockton, Parkdale Village, Exhibition Place     0.000000   
2   Business reply mail Processing Centre, South C...     0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
4                                  Central Bay Street     0.014706   
5                                            Christie     0.000000   
6                                Church and Wellesley     0.026667   
7                      Commerce Court, Victoria Hotel     0.000000   
8                                          Davisville     0.000000   
9                                    Davisville North     0.000000   
10                       Dufferin, Dovercourt Village     0.000000   
11             First Canadian Place, Underground city     0.000000   
12    Forest Hill North & West, Forest Hill Road Park     0.000000   
13                           Garden District, Ryerson     0.000000   
14  Harbourfront East, Union Station, Toronto Islands     0.000000   
15                      High Park, The Junction South     0.000000   
16                     India Bazaar, The Beaches West     0.000000   
17          Kensington Market, Chinatown, Grange Park     0.000000   
18                                      Lawrence Park     0.000000   
19                           Little Portugal, Trinity     0.022222   
20                        Moore Park, Summerhill East     0.000000   
21                  North Toronto West, Lawrence Park     0.055556   
22                             Parkdale, Roncesvalles     0.000000   
23        Queen's Park, Ontario Provincial Government     0.030303   
24                          Regent Park, Harbourfront     0.022727   
25                           Richmond, Adelaide, King     0.000000   
26                                           Rosedale     0.000000   
27                                           Roselawn     0.000000   
28                                 Runnymede, Swansea     0.030303   
29                                     St. James Town     0.000000   
30                        St. James Town, Cabbagetown     0.000000   
31                                     Stn A PO Boxes     0.010417   
32                                    Studio District     0.027027   
33  Summerhill West, Rathnelly, South Hill, Forest...     0.000000   
34                The Annex, North Midtown, Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                       The Danforth West, Riverdale     0.023256   
37           Toronto Dominion Centre, Design Exchange     0.000000   
38                     University of Toronto, Harbord     0.029412   

    Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0.000000   0.0000              0.0000        0.0000   
1            0.000000   0.0000              0.0000        0.0000   
2            0.000000   0.0000              0.0000        0.0000   
3            0.000000   0.0625              0.0625        0.0625   
4            0.000000   0.0000              0.0000        0.0000   
5            0.000000   0.0000              0.0000        0.0000   
6            0.013333   0.0000              0.0000        0.0000   
7            0.000000   0.0000              0.0000        0.0000   
8            0.000000   0.0000              0.0000        0.0000   
9            0.000000   0.0000              0.0000        0.0000   
10           0.000000   0.0000              0.0000        0.0000   
11           0.000000   0.0000              0.0000        0.0000   
12           0.000000   0.0000              0.0000        0.0000   
13           0.000000   0.0000              0.0000        0.0000   
14           0.000000   0.0000              0.0000        0.0000   
15           0.000000   0.0000              0.0000        0.0000   
16           0.000000   0.0000              0.0000        0.0000   
17           0.0

**Confirm the new size**

In [76]:
toronto_grouped.shape

(39, 237)

**Verify results by printing each neighborhood along with the top 5 most common venues**

In [77]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
         venue  freq
0  Coffee Shop  0.09
1     Beer Bar  0.04
2       Bakery  0.04
3   Restaurant  0.04
4  Cheese Shop  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.13
1       Nightclub  0.09
2     Coffee Shop  0.09
3  Breakfast Spot  0.09
4             Gym  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                  venue  freq
0           Pizza Place  0.06
1      Recording Studio  0.06
2            Skate Park  0.06
3  Fast Food Restaurant  0.06
4                  Park  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                 venue  freq
0       Airport Lounge  0.12
1      Airport Service  0.12
2  Rental Car Location  0.06
3      Harbor / Marina  0.06
4     Sculpture Garden  0.06


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.18
1   

**Transform data into Pandas dataframe, in preparation for Clustering and Modeling**

In [78]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [79]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...                  Park   
3  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery            Restaurant           Cheese Shop   
1        Breakfast Spot             Nightclub           Coffee Shop   
2           Pizza Place            Skate Park               Brewery   
3       Airport Service      Sculpture Garden              Boutique   
4                  Café        Sandwich Place    Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0              Beer Bar          Cocktail Bar    Seafood Restaurant   
1         Grocery Store         Burrito Place                Bakery   
2         Burrito Place        Farmers Market  Fast Food Restaurant   
3                 Plane   Rental Car Location           Coffee Shop   
4          Burger Joint      Department Store       Bubble Tea Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Farmers Market        Breakfast Spot               Pharmacy  
1          Intersection    Italian Restaurant             Restaurant  
2            Restaurant               Butcher       Recording Studio  
3         Boat or Ferry       Harbor / Marina                    Bar  
4       Thai Restaurant   Japanese Restaurant            Salad Place

## Cluster Neighborhoods

**I ran the _k_-means, to cluster the neighborhoods, for these different clusters: 4,5,6,7,8,9.**

**I found that 8 clusters resulted in the best partitioning.**

In [139]:
# set number of clusters
kclusters = 8

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([1, 7, 7, 7, 1, 7, 1, 1, 1, 1, 7, 1, 2, 1, 1, 7, 1, 7, 6, 7, 3, 1,
       7, 1, 1, 1, 4, 5, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 7], dtype=int32)

**Create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.**

In [141]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode           Borough                                 Neighborhood  \
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
9         M5B  Downtown Toronto                     Garden District, Ryerson   
15        M5C  Downtown Toronto                               St. James Town   
19        M4E      East Toronto                                  The Beaches   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
2   43.654260 -79.360636               1           Coffee Shop   
4   43.662301 -79.389494               1           Coffee Shop   
9   43.657162 -79.378937               1           Coffee Shop   
15  43.651494 -79.375418               1           Coffee Shop   
19  43.676357 -79.293031               0                   Pub   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
2                   Park                   Pub                Bakery   
4            Yoga Studio                  Café         Smoothie Shop   
9         Clothing Store                  Café       Bubble Tea Shop   
15                  Café          Cocktail Bar            Restaurant   
19                 Trail     Health Food Store         Women's Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
2                   Café               Theater        Breakfast Spot   
4               Beer Bar    Italian Restaurant        Sandwich Place   
9    Japanese Restaurant        Cosmetics Shop    Italian Restaurant   
15             Gastropub              Beer Bar   American Restaurant   
19   Distribution Center          Dessert Shop    Dim Sum Restaurant   

   8th Most Common Venue   9th Most Common Venue     10th Most Common Venue  
2                Brewery                     Spa                 Beer Store  
4             Restaurant     Distribution Center                      Diner  
9       Ramen Restaurant  Furniture / Home Store  Middle Eastern Restaurant  
15        Lingerie Store                   Hotel             Cosmetics Shop  
19                 Diner          Discount Store           Doner Restaurant

**Visualize the resulting clusters**

In [142]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters